Sources/Sentry/SentryProfiler.mm

```diff
@@ -136,7 +136,7 @@
     }
 
     const auto sample = [[SentrySample alloc] init];
-    sample.absoluteTimestamp = SentryCurrentDate.systemTime;
+    sample.absoluteTimestamp = backtrace.absoluteTimestamp;
```

Comment:
I don't think all the changes in the C++ code is necessary -- `SamplingProfiler` and `Backtrace` can stay as they are.

`SentryProfiler` can be initialized with a block with the signature: `const Backtrace & -> uint64_t` (maybe this initializer needs to go in a test-only header, doesn't need to be part of the public API). This block would be called on this line to set `sample.absoluteTimestamp`

When you default initialize this outside of the test, the block should simply return `backtrace.absoluteTimestamp` to preserve the current behavior.

If you need to mock it for the test, the test can pass some other block to the `SentryProfiler` initializer that returns either a fixed value (as you're doing here with `SentryCurrentDate.systemTime`) or an offset added to `backtrace.absoluteTimestamp`

That means that the only changes would need to be in this file and everything else can go away.

Reply:
I don't have a clear idea of what this would look like. If we do this, then we have to wait until after `f` is already finished executing for it to return the timestamp, right? In that case, we aren't getting a timestamp until after the rest of the backtrace has been processed. That will be functionally the same as my initial change, where we just set the sample's timestamp to the value returned by the date provider after processing the backtrace: https://github.com/getsentry/sentry-cocoa/pull/2849/files/446ca92e0bfccf4bf51152bee81c64ac361a5aef..93ce1c2aaadd257108c6f7c1e34233af0588fd89#diff-65ccfc43cdffcd8d20527ee954818c2298714c3790e8d55eb1ab06cee4c7beb9L111. This would be simpler than modifying the callback to return a value.

Maybe I'm just misunderstanding what is being asked for. Would you mind opening a PR into this branch to show me?